src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs

```diff
@@ -14,13 +14,172 @@
+                // unless we know that we are not streaming anymore
+                incomingStream = new WasiInputStream(this, incomingResponse);// passing self ownership, passing incomingResponse ownership
+                response.Content = new StreamContent(incomingStream); // passing incomingStream ownership to SendAsync() caller
+
+                incomingResponse.Dispose();
+
+                return response;
+            }
+            catch (WitException e)
+            {
+                Dispose();
+                throw new HttpRequestException(WasiHttpInterop.ErrorCodeToString((ErrorCode)e.Value), e);
+            }
+            catch (Exception)
+            {
+                Dispose();
+                throw;
+            }
+        }
+
+        private async Task<IncomingResponse> SendRequest(CancellationToken cancellationToken)
+        {
+            try
+            {
+                while (true)
+                {
+                    var response = (Result<Result<IncomingResponse, ErrorCode>, None>?)future!.Get();
+                    if (response.HasValue)
+                    {
+                        var result = response.Value.AsOk;
+
+                        if (result.IsOk)
+                        {
+                            return result.AsOk;
+                        }
+                        else
+                        {
+                            throw new HttpRequestException(WasiHttpInterop.ErrorCodeToString(result.AsErr));
+                        }
+                    }
+                    else
+                    {
+                        await WasiHttpInterop.RegisterWasiPollable(future.Subscribe(), cancellationToken).ConfigureAwait(false);
+                    }
+                }
+            }
+            catch (OperationCanceledException oce)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    Http.CancellationHelper.ThrowIfCancellationRequested(oce, cancellationToken);
+                }
+                throw;
+            }
+        }
+
+        public async Task SendContent(HttpContent? content, OutgoingRequest outgoingRequest, CancellationToken cancellationToken)
+        {
+            if (content is not null)
+            {
+                wasiOutputStream = new WasiOutputStream(outgoingRequest.Body()); // passing body ownership
+                await content.CopyToAsync(wasiOutputStream, cancellationToken).ConfigureAwait(false);
+                wasiOutputStream.Close();
+            }
+        }
+
+        ~WasiRequestWrapper()
+        {
+            Dispose();
+            GC.SuppressFinalize(this);
+        }
+
+        public void Dispose()
+        {
+            if (!isDisposed)
+            {
+                isDisposed = true;
+                wasiOutputStream?.Dispose();
+                incomingStream?.Dispose();
+
+                // TODO why this fails ?
+                // future?.Dispose();
```

Comment:
@dicej Could you please help me to figure this out ? 
It traps with `resource has children` for me.

Reply:
The only kind of child resource a `future-incoming-response` can have is a `pollable` returned from `subscribe`, so it must be that.  Perhaps `WasiEventLoop` is not disposing pollables in all cases before resolving the corresponding `TaskCompletionSource`s?  Might want to add logging here, after the `await` in `SendRequest`, and before resolving the `TaskCompletionSource` in `WasiEventLoop` to see what happens.

Reply:
Exactly